
############################  GENE FITNESS AND ASSOCIATED VARIANCE PER REPLICATE  ############################
############################         PART I: RB-TNSEQ comparative analysis        ############################

# That script processes the raw counts data from the allpoolcounts.tab file generated by the perl script BarSeqTest.pl (Wetmore et al., 2015)
# This script processes 1 biological replicate
# This script produces a final file containing normalized gene fitness and associated fitness variance 
# Many plots are generated along data processing to visually follow data transformation


rm(list=ls())

# Step 1: Packages

library(ggplot2)
library(dplyr)


# Step 2: Sourcing functions required in that script
source("Data_prep_viz10KB.R")
source("Ref_counts_correction.R")
source("Unnorm_gene_fitness.R")
source("Loc_smoothmed_norm.R")
source("Mean_or_mode_normalization.R")

# Step 3: Data import and parameters set up

 # You need to import the table containing the number of counts per barcodes in each condition (allpoolcounts.tab transformed as a csv file - make sure to keep all columns)
 # The first 7 columns of your table should be: barcodes, rbarcode, scaffold, strand, pos, locusId and f, then each column should be a condition ==> THIS IS IMPORTANT THAT THE FIRST 7 COLUMNS ARE NOT COUNTS
 # ALSO COLUMN 8 MUST BE NAMED T0
 # THE CONDITIONS MUST HAVE THE SAME NAME IN EACH REPLICATE
 # You also need to import the genes.GC (as a .txt file) file used to run the perl script TestBarSeq.pl (Wetmore et al., 2015). 

Data_original=read.csv("Raw_Data_Example.csv")  # import your allpoolcounts filr
head(Data_original,n=5)

genes.tab <- readr::read_delim("genesExample.GC.txt", 
                               "\t", escape_double = FALSE, trim_ws = TRUE)  # import your gene.GC file
head(genes.tab, n=5)

 # Depending on your genome annotations, the locus_Id may be numerics or characters, you have to specify it with org_locId
 # Set up the number of analyzed conditions (including T0) with cdnb
 # Set up the chromosome scaffold (for plot purposes only)
 # Indicate the locusId of the gene you will use as a reference to normalize counts

org_locId="Num"   # "Num" if numeric , "Char" if characters
cdnb=9 
scaffoldX=7023  
ref=14649 # here you write the locusId of your reference gene CAREFUL, depending on the locusId Type, it might be a numeric  or character 



# Step 4: Data pre-visualization

 # Before processing the data we represent, for each condition (T0 included) the number of counts per 10kB 

Data=Data_original

Data_prep=Data_prep_viz10KB(Data,cdnb,scaffoldX)    # format the data for vizualisation as counts per 10kB

dat_format <- data.frame(Data_prep[ncol(Data_prep)], stack(Data_prep[1:cdnb]))
colnames(dat_format)=c("Rank","Counts","Cdt")

plot1=ggplot(dat_format, aes(x=Rank,y=Counts,col=Cdt)) + geom_point(shape=19) + theme_light() + ggtitle("Counts per 10kb interval") +
  facet_grid(Cdt ~ .) + labs(x = "Chromosome position (10kB)", y="number of counts per 10kb interval")
#Note:if a couple of outliers points prevent from accuretely observing the number of counts per 10kB, you can change y=Counts to y=log10(Counts)


# Step 5: Count correction before fitness calculation

 # A pseudocount of .1 is added to each count to avaoid counts of 0
 # Insertions mutants that are outside of the ORF (f<0.1 and f>0.9) are filtered out
 # Insertion mutants that do not pass the T0 count threshold (before correction relative to reference ; Counts<3.1) are filtered out
 # Counts are transformed to be expressed as relative to the counts of the reference gene
 # Data are visualized again after correction

Data_ref_corrected=Ref_counts_correction(Data, ref, cdnb)

Dat_viz=Data_prep_viz10KB(Data_ref_corrected,cdnb,scaffoldX)
dat_format <- data.frame(Dat_viz[ncol(Dat_viz)], stack(Dat_viz[1:cdnb]))
colnames(dat_format)=c("Rank","Counts","Cdt")

plot2=ggplot(dat_format, aes(x=Rank,y=Counts,col=Cdt)) + geom_point(shape=19) + theme_light() + ggtitle("Corrected counts per 10kb interval") +
  facet_grid(Cdt ~ .) + labs(x = "Chromosome position (10kB)", y="number of counts per 10kb interval after Ref correction")

#Note:if a couple of outliers points prevent from accuretely observing the number of counts per 10kB, you can change y=Counts to y=log10(Counts)



# Step 6: Calculation of un-normalized gene fitness
 # Calculation of the strain fitness (fitness for each insertion with central insertion as the log2 of the ratio of the corrected counts in the condition and the corrected counts in the T0)
 # Calculation of the un-normalized gene fitness along with its variance (average of all the insertion mutants fitness in that gene)
 # Vizualization of the un-normalized fitness values (along the chromose (plot3) or as a distribution (plot4))

Data_for_fit=Data_ref_corrected
cdnbF=(cdnb-1)   # number of conditions for which we will get a fitness values (=everything but T0)

Unnorm_values=Unnorm_gene_fitness(Data_for_fit,cdnbF,genes.tab,org_locId)  # This function returns a list containing 3 tables: "Strain fitness": strain fitness values, "Gene_unnorm_fitness": unnormalized gene fitness values and "Gene_fitness_variance": variance assocaited with unnormalized gene fitness values 

Table_GeneFitness_MEAN=Unnorm_values[[2]]  # We extract the table with the unnormalized fitness values for data vizualisation

 #We format the data to vizualise them 
dat_format <- data.frame(Table_GeneFitness_MEAN[1], Table_GeneFitness_MEAN[(ncol(Table_GeneFitness_MEAN)-1):ncol(Table_GeneFitness_MEAN)], stack(Table_GeneFitness_MEAN[2:(2+cdnbF-1)]))
colnames(dat_format)=c("locusId","sysName","begin","RawFitness","Cdt")

plot3=ggplot(dat_format, aes(x=begin,y=RawFitness,col=Cdt)) + geom_point(shape=19) + theme_light() + ggtitle("Un-normalized gene fitness") +
  facet_grid(Cdt ~ .) + labs(x = "Chromosome position", y="Gene un-normalized fitness")

plot4=ggplot(dat_format, aes(x=RawFitness,col=Cdt)) + geom_density(aes(fill=Cdt)) + theme_light()+ ggtitle("Distribution of gene un-normalized fitness values") +
  facet_grid(Cdt ~ .) + labs(x = "Gene un-normalized fitness", y="Density")




# Step 7: Gene fitness normalization
 # Normalize gene fitness based on the position of the gene on the chromosome (smooth median normalization ; Wetmore et al., 2015)
 # Normalize gene fitness by mean or mode of the fitness distribution (for each condition individually) mean of the distribution or the mode
 # Data visualization like Step 6

Data_to_norm=Table_GeneFitness_MEAN
Data_norm_loc=Loc_smoothmed_norm(Data_to_norm,genes.tab,cdnbF,org_locId)  # we obtain the gene fitness values normalized by the smoothed median ==> normalization for gene location

Data_to_norm2=Data_norm_loc   # Note: in this table, everything is a character ==> no numeric ==> need to transform values in numeric if you want to plot at that stage

  # Normalization around mean
Data_norm_mean_OP=Norm_around_mean(Data_to_norm2,cdnbF) # we center around the mean ==> function output is a list with [1] dataframe containing normalized gene fitness per gene and [2] the mean fitness value used for normalization
Data_norm_mean=Data_norm_mean_OP[[1]]

  # Normalization around mode
Data_norm_mode=Norm_around_mode(Data_to_norm2) 


 # plots after normalization by location and centerd around the mean
dat_format <- data.frame(Data_norm_mean[1:4], stack(Data_norm_mean[5:ncol(Data_norm_mean)]))
colnames(dat_format)=c("locusId","sysName","begin","scaffold","NormMeanFitness","Cdt")
dat_format$begin=as.numeric(dat_format$begin)

plot5=ggplot(dat_format, aes(x=begin, y=NormMeanFitness, col=Cdt)) + geom_point(shape=19) + theme_light() + ggtitle("Normalized gene fitness values (centered around the mean)") +
  facet_grid(Cdt ~ .) + labs(x = "Chromosome position (10kB)", y="Chrom + Mean normalized gene fitness")

plot6=ggplot(dat_format, aes(x=NormMeanFitness, col=Cdt)) + geom_density(aes(fill=Cdt)) + theme_light() + ggtitle("Distribution of normalized gene fitness values (centered around the mean)") +
  facet_grid(Cdt ~ .) + labs(y = "Density", x="Chrom + Mean normalized gene fitness")



 # plots after normalization by location and centerd around the mode
dat_format <- data.frame(Data_norm_mode[1:4], stack(Data_norm_mode[5:ncol(Data_norm_mode)]))
colnames(dat_format)=c("locusId","sysName","begin","scaffold","NormModeFitness","Cdt")
dat_format$begin=as.numeric(dat_format$begin)

plot7=ggplot(dat_format, aes(x=begin, y=NormModeFitness, col=Cdt)) + geom_point(shape=19) + theme_light() + ggtitle("Normalized gene fitness values (centered around the mode)") +
  facet_grid(Cdt ~ .) + labs(x = "Chromosome position (10kB)", y="Chrom + Mode normalized gene fitness")

plot8=ggplot(dat_format, aes(x=NormModeFitness, col=Cdt)) + geom_density(aes(fill=Cdt)) + theme_light() + ggtitle("Distribution of ormalized gene fitness values (centered around the mode)") +
  facet_grid(Cdt ~ .) + labs(y = "Density", x="Chrom + Mode normalized gene fitness")

save(plot7,file= "plot7_NormFitnessmode_Chromosome_Rep3.pdf")
save(plot8,file= "plot8_NormFitnessmode_Density_Rep3.pdf")

# Step 8: Data formating and save

  # Data normalized fitness values by mode
Data_Fitness_mode=data.frame(Data_norm_mode[1:4], stack(Data_norm_mode[5:ncol(Data_norm_mode)]))
colnames(Data_Fitness_mode)=c("locusId","sysName","begin","scaffold","Fitness_byMode","Cdt")

  # Data normalized fitness values by mean
Data_Fitness_mean=data.frame(Data_norm_mean[1:4], stack(Data_norm_mean[5:ncol(Data_norm_mean)]))
colnames(Data_Fitness_mean)=c("locusId","sysName","begin","scaffold","Fitness_byMean","Cdt")

Table_GeneFitness_VAR=Unnorm_values[[3]]
Data_Fitness_VAR=data.frame(Table_GeneFitness_VAR[1], stack(Table_GeneFitness_VAR[2:ncol(Table_GeneFitness_VAR)]))
colnames(Data_Fitness_VAR)=c("locusId","Fitness_Variance","Cdt")

All_data_Replicate=left_join(Data_Fitness_mean,Data_Fitness_mode,by=c("locusId","sysName","begin","scaffold","Cdt"))

if (org_locId=="Num"){
  All_data_Replicate$locusId=as.numeric(All_data_Replicate$locusId)
}

All_data_Replicate=left_join(All_data_Replicate,Data_Fitness_VAR,by=c("locusId","Cdt"))
All_data_Replicate=All_data_Replicate[c(1,2,3,4,5,7,8,6)]

#save(All_data_Replicate, Data_norm_mean_OP, Data_norm_mode, Data_norm_mean,
     #Unnorm_values, Data_ref_corrected, Data_original, genes.tab,
     #plot1, plot2, plot3, plot4, plot5, plot6, plot7, plot8,
     #file="EcoliTest_Rep1.Rdata")

save(All_data_Replicate, Data_norm_mean_OP, Data_norm_mode, Data_norm_mean,
     Unnorm_values, Data_ref_corrected, Data_original, genes.tab,
     file="")









